runtime-sdk/src/modules/consensus_accounts/mod.rs

```diff
@@ -0,0 +1,383 @@
+pub struct Module<Accounts: modules::accounts::API, Consensus: modules::consensus::API> {
+    _accounts: std::marker::PhantomData<Accounts>,
+    _consensus: std::marker::PhantomData<Consensus>,
+}
+
+const CONSENSUS_TRANSFER_HANDLER: &str = "consensus.TransferFromRuntime";
+const CONSENSUS_WITHDRAW_HANDLER: &str = "consensus.WithdrawIntoRuntime";
+
+impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API> API
+    for Module<Accounts, Consensus>
+{
+    fn deposit<C: Context>(
+        ctx: &mut C,
+        from: Address,
+        amount: token::BaseUnits,
+    ) -> Result<(), Error> {
+        if ctx.is_check_only() {
+            return Ok(());
+        }
+
+        // XXX: could check consensus state if allowance for the runtime account
+        // exists, but consensus state could be outdated since last block, so
+        // just try to withdraw.
+
+        // Do withdraw from the consensus account and update the account state if
+        // successful.
+        Consensus::withdraw(
+            ctx,
+            from,
+            &amount,
+            MessageEventHookInvocation::new(
+                CONSENSUS_WITHDRAW_HANDLER.to_string(),
+                types::ConsensusWithdrawContext {
+                    address: from,
+                    amount: amount.clone(),
+                },
+            ),
+        )?;
+
+        Ok(())
+    }
+
+    fn withdraw<C: Context>(
+        ctx: &mut C,
+        to: Address,
+        amount: token::BaseUnits,
+    ) -> Result<(), Error> {
+        if ctx.is_check_only() {
+            return Ok(());
+        }
+
+        // Check internal store if account has enough balance to withdraw.
+        let balances =
+            Accounts::get_balances(ctx.runtime_state(), to).map_err(|_| Error::InvalidArgument)?;
+        let balance = balances
+            .balances
+            .get(amount.denomination())
+            .ok_or(Error::InvalidArgument)?;
+        if balance < amount.amount() {
+            return Err(Error::InsufficientWithdrawBalance);
+        }
+
+        // Transfer out of runtime account and update the account state if successful.
+        Consensus::transfer(
+            ctx,
+            to,
+            &amount,
+            MessageEventHookInvocation::new(
+                CONSENSUS_TRANSFER_HANDLER.to_string(),
+                types::ConsensusTransferContext {
+                    address: to,
+                    amount: amount.clone(),
+                },
+            ),
+        )?;
+
+        Ok(())
+    }
+}
+
+impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API>
+    Module<Accounts, Consensus>
+{
+    /// Deposit in the runtime.
+    fn tx_deposit(ctx: &mut TxContext<'_, '_>, body: types::Deposit) -> Result<(), Error> {
+        let signer = &ctx
+            .tx_auth_info()
+            .expect("should be called with a transaction ctx")
+            .signer_info[0];
+        let address = Address::from_pk(&signer.public_key);
```

Comment:
This should make sure that the address is compatible with the consensus layer (e.g., is derived from an Ed25519 key).